diff --git a/esm/pluggableElementTypes/models/components/SaveTrackData.js b/esm/pluggableElementTypes/models/components/SaveTrackData.js
index c2122713d61f844501864e1fe2acf90255c14eea..137fb13bc625d47a0fda43c05d67d1487bfb1f9a 100644
--- a/esm/pluggableElementTypes/models/components/SaveTrackData.js
+++ b/esm/pluggableElementTypes/models/components/SaveTrackData.js
@@ -2,7 +2,7 @@ import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
 import { useEffect, useMemo, useState } from 'react';
 import ContentCopyIcon from '@mui/icons-material/ContentCopy';
 import GetAppIcon from '@mui/icons-material/GetApp';
-import HelpOutlineIcon from '@mui/icons-material/HelpOutline';
+import HelpOutlineIcon from '@mui/icons-material/HelpOutlined';
 import { Button, DialogActions, DialogContent, DialogContentText, DialogTitle, FormControl, FormControlLabel, FormLabel, IconButton, Radio, RadioGroup, TextField, } from '@mui/material';
 import copy from 'copy-to-clipboard';
 import { saveAs } from 'file-saver-es';
diff --git a/esm/ui/CascadingMenuHelpIconButton.js b/esm/ui/CascadingMenuHelpIconButton.js
index cca45dc589f7ac587099f24fb5451105f718df8d..fbdeb364dff1fefd46527cdae8380c41b0b5c722 100644
--- a/esm/ui/CascadingMenuHelpIconButton.js
+++ b/esm/ui/CascadingMenuHelpIconButton.js
@@ -1,6 +1,6 @@
 import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
 import { Suspense, useState } from 'react';
-import HelpOutline from '@mui/icons-material/HelpOutline';
+import HelpOutline from '@mui/icons-material/HelpOutlined';
 import { IconButton } from '@mui/material';
 import CascadingMenuHelpDialog from "./CascadingMenuHelpDialog.js";
 export default function CascadingMenuHelpIconButton({ helpText, label, }) {
